src/vcpkg/visualstudio.cpp

```diff
@@ -138,12 +218,184 @@ namespace vcpkg::VisualStudio
                         Checks::unreachable(VCPKG_LINE_INFO);
                 }
 
+                const auto& vs_version =
+                    Strings::find_exactly_one_enclosed(instance, "<installationVersion>", "</installationVersion>")
+                        .to_string();
+
+                CStringView x86_64_toolset;
+                // For Visual Studio 2015
+                if (0 == vs_version.compare(0, 3, "14."))
+                {
+                    x86_64_toolset = COMPONENT_VCTOOLS_2015;
+                }
+                // For Visual Studio 2017 or later
+                else if (0 == vs_version.compare(0, 3, "15.") || 0 == vs_version.compare(0, 3, "16.") ||
+                         0 == vs_version.compare(0, 3, "17."))
+                {
+                    x86_64_toolset = COMPONENT_VCTOOLS_2017_OR_LATER;
+                }
+                // For unknown Visual Studio version
+                else
+                {
+                    print2("Unknown Visual Studio version: %s", vs_version);
+                }
+
+                Components_Support support = Installed_Component::none;
+                if (x86_64_toolset != "")
+                {
+                    // VC++
+                    const auto native_desktop_output =
+                        cmd_execute_and_capture_output(Command(vswhere_exe)
```

Comment:
We running vswhere.exe individually this many times (on the happy path) is likely to be a serious performance problem. We may want to use the query interface directly: https://www.nuget.org/packages/Microsoft.VisualStudio.Setup.Configuration.Native/.

Otherwise, we should instead use an approach of:
1. First, run a single vswhere call with _all_ of the `-requires` statements.
2. Only if that first call fails, then fall back to individual queries to provide better experiences for customers

Reply:
@ras0219-msft Another questions:
We probably don't need to make sure these components are all installed. 
For example: you don't need to install arm/arm64 when building for x86/x64.
This means that we need to get the architecture used here and do it.

However, we can only get the triplet name here. Judging by the triplet name is not necessarily a good idea.

Reply:
> Judging by the triplet name is not necessarily a good idea.

Yes, absolutely.

> However, we can only get the triplet name here.

We should switch on the value of `VCPKG_TARGET_ARCHITECTURE`, which is part of the `PreBuildInfo` and should be available above this function in the call stack.

Reply:
@ras0219-msft 
If we look for the **required** components based on `target_architecture`, we won't find all the components. 
This means that `The available toolchain combinations are {list}` in the above message cannot be implemented.

Reply:
Also, I found that uwp support is not checked in function `find_toolset_instances_preferred_first`: now we only check for x86/x64/arm/arm64.

Reply:
https://github.com/microsoft/vcpkg-tool/blob/e65f78439a43641a3ef2c5f060510de142caa7a3/src/vcpkg/visualstudio.cpp#L234-L250

I wonder if the content here can be replaced now that we have verified the installed components.
Also, no uwp related stuff is set here. Is it appropriate if I add uwp to the enum `CPU`?